api/v1alpha1/database_types.go

```diff
@@ -10,52 +10,101 @@ type DatabaseSpec struct {
 	// Number of nodes (pods) in the cluster
 	// +required
 	Nodes int32 `json:"nodes"`
+
 	// (Optional) Storage services parameter overrides
 	// Default: (not specified)
 	// +optional
 	Service DatabaseServices `json:"service,omitempty"`
+
 	// YDB Storage cluster reference
 	// +required
 	StorageClusterRef StorageRef `json:"storageClusterRef"`
+
 	// (Optional) Name of the root storage domain
 	// Default: root
 	// +kubebuilder:validation:Pattern:=[a-zA-Z0-9]([-_a-zA-Z0-9]*[a-zA-Z0-9])?
 	// +kubebuilder:validation:MaxLength:=63
 	// +kubebuilder:default:="root"
 	// +optional
 	Domain string `json:"domain"`
-	// (Optional) Database container resource limits. Any container limits
-	// can be specified.
-	// Default: (not specified)
+
+	// (Optional) Database storage and compute resources
+	// +optional
+	Resources *DatabaseResources `json:"resources,omitempty"`  // TODO: Add validation webhook: some resources must be specified
+
+	// (Optional) Shared resources can be used by serverless databases.
 	// +optional
-	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
+	SharedResources *DatabaseResources `json:"sharedResources,omitempty"`
+
+	// (Optional) If specified, created database will be "serverless".
+	// +optional
+	ServerlessResources *ServerlessDatabaseResources `json:"serverlessResources,omitempty"`
```

Comment:
Can we collapse those three types to one property? And use an `Enum` for type definition?

Reply:
I thought of this option first, but it complicates validation logic and user experience while filling in custom resource fields. User must remember which options to fill when he chooses specific database type.

Also current solution is better in terms of natural language and has better descriptive properties. Looking at custom resource you can tell that this database has shared resources. And another one has serverless resources.

This is also debatable and really hard to make correct choice.